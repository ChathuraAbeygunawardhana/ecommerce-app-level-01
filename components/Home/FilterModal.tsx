import React from 'react';
import {
  StyleSheet,
  View,
  Text,
  TouchableOpacity,
  Image,
  TextInput,
} from 'react-native';

const nikeBlackLogo = 'https://i.ibb.co/vHp5FV7/nikeblacklogo.png';
const nikeWhiteLogo = 'https://i.ibb.co/C17pnkw/nikewhitelogo.png';
const pumaBlackLogo = 'https://i.ibb.co/HT5F046/pumablacklogo.png';
const pumaWhiteLogo = 'https://i.ibb.co/wzVBnzj/pumawhitelogo.png';

interface FilterModalProps {
  uniqueColors: string[];
  brand: string;
  setBrand: (brand: string) => void;
  color: string;
  setColor: (color: string) => void;
  minPrice: string;
  setMinPrice: (price: string) => void;
  maxPrice: string;
  setMaxPrice: (price: string) => void;
  applyFilters: () => void;
  clearFilters: () => void;
}

const FilterModal: React.FC<FilterModalProps> = ({
  uniqueColors,
  brand,
  setBrand,
  color,
  setColor,
  minPrice,
  setMinPrice,
  maxPrice,
  setMaxPrice,
  applyFilters,
  clearFilters,
}) => (
  <View style={styles.modalOverlay}>
    <View style={styles.bottomSheet}>
      <Text style={styles.modalTitle}>Filter Products</Text>
      <View style={styles.brandContainer}>
        <TouchableOpacity
          style={[
            styles.brandButton,
            brand === 'Nike' && styles.selectedBrandButton,
          ]}
          onPress={() => setBrand('Nike')}
        >
          <Image
            source={{
              uri: brand === 'Nike' ? nikeWhiteLogo : nikeBlackLogo,
            }}
            style={styles.brandImage}
          />
          <Text
            style={[
              styles.brandButtonText,
              brand === 'Nike' && styles.selectedBrandButtonText,
            ]}
          >
            Nike
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[
            styles.brandButton,
            brand === 'Puma' && styles.selectedBrandButton,
          ]}
          onPress={() => setBrand('Puma')}
        >
          <Image
            source={{
              uri: brand === 'Puma' ? pumaWhiteLogo : pumaBlackLogo,
            }}
            style={styles.brandImage}
          />
          <Text
            style={[
              styles.brandButtonText,
              brand === 'Puma' && styles.selectedBrandButtonText,
            ]}
          >
            Puma
          </Text>
        </TouchableOpacity>
      </View>
      <View style={styles.colorContainer}>
        {uniqueColors.map((colorOption) => (
          <TouchableOpacity
            key={colorOption}
            style={[
              styles.colorButton,
              color === colorOption && styles.selectedColorButton,
            ]}
            onPress={() => setColor(colorOption)}
          >
            <Text
              style={[
                styles.colorButtonText,
                color === colorOption && styles.selectedColorButtonText,
              ]}
            >
              {colorOption}
            </Text>
          </TouchableOpacity>
        ))}
      </View>
      <TextInput
        style={styles.input}
        placeholder="Min Price"
        keyboardType="numeric"
        value={minPrice}
        onChangeText={setMinPrice}
      />
      <TextInput
        style={styles.input}
        placeholder="Max Price"
        keyboardType="numeric"
        value={maxPrice}
        onChangeText={setMaxPrice}
      />
      <View style={styles.buttonContainer}>
        <TouchableOpacity style={[styles.button, styles.applyButton]} onPress={applyFilters}>
          <Text style={styles.buttonText}>Apply Filters</Text>
        </TouchableOpacity>
        <TouchableOpacity style={[styles.button, styles.clearButton]} onPress={clearFilters}>
          <Text style={styles.buttonText}>Clear</Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>
);

export default FilterModal;

const styles = StyleSheet.create({
  modalOverlay: {
    flex: 1,
    justifyContent: 'flex-end',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  bottomSheet: {
    backgroundColor: 'white',
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,
    padding: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  modalTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 15,
    textAlign: 'center',
  },
  brandContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 10,
  },
  brandButton: {
    flex: 1,
    padding: 10,
    marginHorizontal: 5,
    backgroundColor: 'white',
    borderRadius: 5,
    alignItems: 'center',
    flexDirection: 'row',
    borderWidth: 1,
    borderColor: 'black',
  },
  selectedBrandButton: {
    backgroundColor: 'black',
    borderColor: 'black',
  },
  brandButtonText: {
    color: 'black',
    fontWeight: 'bold',
  },
  selectedBrandButtonText: {
    color: 'white',
  },
  colorContainer: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginBottom: 10,
  },
  colorButton: {
    padding: 10,
    margin: 5,
    borderWidth: 1,
    borderColor: 'black',
    backgroundColor: 'white',
    borderRadius: 5,
    alignItems: 'center',
  },
  selectedColorButton: {
    backgroundColor: 'black',
  },
  colorButtonText: {
    color: 'black',
    fontWeight: 'bold',
  },
  selectedColorButtonText: {
    color: 'white',
  },
  input: {
    height: 40,
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 5,
    marginBottom: 10,
    paddingHorizontal: 10,
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 10,
  },
  button: {
    width: '48%',
    alignItems: 'center',
    padding: 10,
    borderRadius: 5,
  },
  applyButton: {
    backgroundColor: 'black',
  },
  clearButton: {
    backgroundColor: 'black',
  },
  buttonText: {
    color: 'white',
    fontWeight: 'bold',
  },
  brandImage: {
    width: 20,
    height: 20,
    marginRight: 5,
  },
});
