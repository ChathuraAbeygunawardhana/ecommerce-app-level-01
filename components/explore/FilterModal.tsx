import React, { useState, useEffect } from 'react';
import { View, Text, TextInput, TouchableOpacity, Modal, StyleSheet } from 'react-native';
import { Colors } from '../../constants/Colors';
import { useTheme } from '../../contexts/ThemeContext';
import sampleData from '../../assets/sample.json';

interface FilterModalProps {
  modalVisible: boolean;
  setModalVisible: (visible: boolean) => void;
  minPrice: string;
  setMinPrice: (price: string) => void;//TODO: change to number
  maxPrice: string; //TODO: change to number
  setMaxPrice: (price: string) => void; //TODO: change to number
  selectedBrands: string[];
  toggleBrand: (brand: string) => void;
  clearFilters: () => void;
  applyFilters: () => void;
}

const FilterModal: React.FC<FilterModalProps> = ({
  modalVisible,
  setModalVisible,
  minPrice,
  setMinPrice,
  maxPrice,
  setMaxPrice,
  selectedBrands,
  toggleBrand,
  clearFilters,
  applyFilters,
}) => {
  const { theme } = useTheme();
  const currentColors = Colors[theme as 'light' | 'dark'];
  const currentSpecialGrey = theme === 'light' ? Colors.light.specialGrey : Colors.dark.specialGrey;
  const brands = ['Nike', 'Adidas', 'PUMA', 'NB'];

  const [isPriceValid, setIsPriceValid] = useState(true);

  useEffect(() => {
    const min = minPrice ? parseFloat(minPrice) : 0;
    const max = maxPrice ? parseFloat(maxPrice) : Infinity;

    setIsPriceValid(min <= max);
  }, [minPrice, maxPrice]);

  const handleApplyFilters = () => {
    if (!isPriceValid) return;
    applyFilters();
  };

  return (
    <Modal
      animationType="slide"
      transparent={true}
      visible={modalVisible}
      onRequestClose={() => setModalVisible(false)}
    >
      <TouchableOpacity
        style={styles.modalOverlay}
        activeOpacity={1}
        onPressOut={() => setModalVisible(false)}
      >
        <View
          style={[
            styles.bottomSheet,
            { backgroundColor: currentColors.background_02 },
          ]}
        >
          <TouchableOpacity activeOpacity={1}>
            <Text
              style={[
                styles.modalTitle,
                { color: currentColors.text, fontSize: 22 },//TODO: font size can move to modalTitle styles
              ]}
            >
              Filters
            </Text>

            <Text
              style={[
                styles.modalText,
                { color: currentColors.text, textAlign: 'left' }, //TODO: text align can move to modalText styles
              ]}
            >
              Price Range
            </Text>
            <View style={styles.priceContainer}>
              <TextInput
                style={[
                  styles.input,
                  {
                    color: currentColors.text,
                    borderColor: currentColors.grey,
                    borderRadius: 60,
                  },
                ]}
                placeholder="Min Price"
                placeholderTextColor={Colors.grey}
                keyboardType="numeric"
                value={minPrice}
                onChangeText={setMinPrice}
              />
              <TextInput
                style={[
                  styles.input,
                  {
                    color: currentColors.text,
                    borderColor: currentColors.grey,
                    borderRadius: 60,
                  },
                ]}
                placeholder="Max Price"
                placeholderTextColor={Colors.grey}
                keyboardType="numeric"
                value={maxPrice}
                onChangeText={setMaxPrice}
              />
            </View>
            {!isPriceValid && (
              <Text style={{ color: 'red', textAlign: 'center' }}>
                Min value should be less than max value
              </Text>
            )}

            <Text
              style={[
                styles.modalText,
                { color: currentColors.text, textAlign: 'left' },
              ]}
            >
              Brands
            </Text>
            <View style={styles.brandContainer}>
              {brands.map((brand) => (
                <TouchableOpacity
                  key={brand}
                  style={[
                    styles.brandButton,
                    {
                      borderRadius: 60,
                      backgroundColor: currentSpecialGrey,
                      borderWidth: 0,
                    },
                    selectedBrands.includes(brand) && {
                      backgroundColor: Colors.lightBlue,
                    },
                  ]}
                  onPress={() => toggleBrand(brand)}
                >
                  <Text
                    style={[
                      styles.brandButtonText,
                      { color: Colors.grey },
                      selectedBrands.includes(brand) && { color: Colors.white },
                    ]}
                  >
                    {brand}
                  </Text>
                </TouchableOpacity>
              ))}
            </View>

            <View style={styles.buttonContainer}>
              <TouchableOpacity
                style={[
                  styles.button,
                  styles.clearButton,
                  {
                    backgroundColor: currentColors.lightBlue,
                    borderRadius: 60,
                  },
                ]}
                onPress={clearFilters}
              >
                <Text style={styles.buttonText}>Clear</Text>
              </TouchableOpacity>
              <TouchableOpacity
                style={[
                  styles.button,
                  styles.applyButton,
                  {
                    backgroundColor: currentColors.lightBlue,
                    borderRadius: 60,
                  },
                ]}
                onPress={handleApplyFilters}
              >
                <Text style={styles.buttonText}>Apply</Text>
              </TouchableOpacity>
            </View>
          </TouchableOpacity>
        </View>
      </TouchableOpacity>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalOverlay: {
    flex: 1,
    justifyContent: 'flex-end',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  bottomSheet: {
    backgroundColor: 'white',
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,
    padding: 20,
    paddingHorizontal: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  modalText: {
    marginBottom: 15,
    textAlign: 'center',
  },
  modalTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 15,
    textAlign: 'center',
  },
  priceContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  input: {
    flex: 1,
    height: 40,
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 5,
    marginBottom: 10,
    paddingHorizontal: 10,
    marginHorizontal: 5,
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 20,
  },
  button: {
    width: '48%',
    alignItems: 'center',
    padding: 10,
    borderRadius: 5,
  },
  applyButton: {
    backgroundColor: 'black',
  },
  clearButton: {
    backgroundColor: 'black',
  },
  buttonText: {
    color: 'white',
  },
  brandContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 10,
  },
  brandButton: {
    flex: 1,
    padding: 10,
    marginHorizontal: 5,
    backgroundColor: 'white',
    borderRadius: 10,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    borderWidth: 1,
    borderColor: 'black',
  },
  selectedBrandButton: {
    backgroundColor: 'black',
    borderColor: 'black',
  },
  brandButtonText: {
    color: 'black',
    fontWeight: 'bold',
  },
  selectedBrandButtonText: {
    color: 'white',
  },
});

export default FilterModal;
