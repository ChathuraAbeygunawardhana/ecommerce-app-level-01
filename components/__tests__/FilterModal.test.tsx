import React from 'react';
import renderer from 'react-test-renderer';
import FilterModal from '../Home/FilterModal';

describe('FilterModal', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <FilterModal
        uniqueColors={['Red', 'Blue']}
        brand="Nike"
        setBrand={() => {}}
        color="Red"
        setColor={() => {}}
        minPrice="0"
        setMinPrice={() => {}}
        maxPrice="100"
        setMaxPrice={() => {}}
        applyFilters={() => {}}
        clearFilters={() => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});