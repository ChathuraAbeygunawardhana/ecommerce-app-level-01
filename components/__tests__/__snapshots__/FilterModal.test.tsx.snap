// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterModal renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(0, 0, 0, 0.5)",
      "flex": 1,
      "justifyContent": "flex-end",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderTopLeftRadius": 20,
        "borderTopRightRadius": 20,
        "elevation": 5,
        "padding": 20,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 4,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 15,
          "textAlign": "center",
        }
      }
    >
      Filter Products
    </Text>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "black",
            "borderColor": "black",
            "borderRadius": 5,
            "borderWidth": 1,
            "flex": 1,
            "flexDirection": "row",
            "marginHorizontal": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            {
              "uri": "https://i.ibb.co/C17pnkw/nikewhitelogo.png",
            }
          }
          style={
            {
              "height": 20,
              "marginRight": 5,
              "width": 20,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "black",
                "fontWeight": "bold",
              },
              {
                "color": "white",
              },
            ]
          }
        >
          Nike
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 5,
            "borderWidth": 1,
            "flex": 1,
            "flexDirection": "row",
            "marginHorizontal": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            {
              "uri": "https://i.ibb.co/HT5F046/pumablacklogo.png",
            }
          }
          style={
            {
              "height": 20,
              "marginRight": 5,
              "width": 20,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "black",
                "fontWeight": "bold",
              },
              false,
            ]
          }
        >
          Puma
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "marginBottom": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "black",
            "borderColor": "black",
            "borderRadius": 5,
            "borderWidth": 1,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "black",
                "fontWeight": "bold",
              },
              {
                "color": "white",
              },
            ]
          }
        >
          Red
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 5,
            "borderWidth": 1,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "black",
                "fontWeight": "bold",
              },
              false,
            ]
          }
        >
          Blue
        </Text>
      </View>
    </View>
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="Min Price"
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 40,
          "marginBottom": 10,
          "paddingHorizontal": 10,
        }
      }
      value="0"
    />
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="Max Price"
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 40,
          "marginBottom": 10,
          "paddingHorizontal": 10,
        }
      }
      value="100"
    />
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "black",
            "borderRadius": 5,
            "opacity": 1,
            "padding": 10,
            "width": "48%",
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontWeight": "bold",
            }
          }
        >
          Apply Filters
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "black",
            "borderRadius": 5,
            "opacity": 1,
            "padding": 10,
            "width": "48%",
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontWeight": "bold",
            }
          }
        >
          Clear
        </Text>
      </View>
    </View>
  </View>
</View>
`;
